modules/core/04-channel/keeper/upgrade_test.go

```diff
@@ -487,6 +487,167 @@ func (suite *KeeperTestSuite) TestChanUpgradeAck() {
 	}
 }
 
+func (suite *KeeperTestSuite) TestChanUpgradeOpen() {
+	var path *ibctesting.Path
+	testCases := []struct {
+		name     string
+		malleate func()
+		expError error
+	}{
+		{
+			"success, counterparty in TRYUPGRADE",
+			func() {
+				err := path.EndpointA.ChanUpgradeInit()
+				suite.Require().NoError(err)
+
+				err = path.EndpointB.ChanUpgradeTry()
+				suite.Require().NoError(err)
+
+				err = path.EndpointA.ChanUpgradeAck()
+				suite.Require().NoError(err)
+
+				channelB := path.EndpointB.GetChannel()
+				channelB.FlushStatus = types.FLUSHCOMPLETE
+				path.EndpointB.SetChannel(channelB)
+
+				channelA := path.EndpointA.GetChannel()
+				channelA.FlushStatus = types.FLUSHCOMPLETE
+				path.EndpointA.SetChannel(channelA)
+
+				suite.coordinator.CommitBlock(suite.chainA, suite.chainB)
+				suite.Require().NoError(path.EndpointA.UpdateClient())
+			},
+			nil,
+		},
+		{
+			"success, counterparty in ACKUPGRADE",
+			func() {
+				err := path.EndpointB.ChanUpgradeInit()
+				suite.Require().NoError(err)
+
+				err = path.EndpointA.ChanUpgradeTry()
+				suite.Require().NoError(err)
+
+				err = path.EndpointB.ChanUpgradeAck()
+				suite.Require().NoError(err)
+
+				channelB := path.EndpointB.GetChannel()
+				channelB.FlushStatus = types.FLUSHCOMPLETE
+				path.EndpointB.SetChannel(channelB)
+
+				channelA := path.EndpointA.GetChannel()
+				channelA.FlushStatus = types.FLUSHCOMPLETE
+				path.EndpointA.SetChannel(channelA)
+
+				suite.coordinator.CommitBlock(suite.chainA, suite.chainB)
+				suite.Require().NoError(path.EndpointA.UpdateClient())
+			},
+			nil,
+		},
+		/*
+			{
+				// TODO: Pending implementation of open handling on msg_server
+				"success, counterparty in OPEN",
+				func() {},
+				nil,
+			},
+		*/
+		{
+			"channel not found",
+			func() {
+				path.EndpointA.ChannelConfig.PortID = ibctesting.InvalidID
+			},
+			types.ErrChannelNotFound,
+		},
+		{
+			"channel state is not in TRYUPGRADE or ACKUPGRADE",
+			func() {
+				channel := path.EndpointA.GetChannel()
+				channel.State = types.OPEN
+				path.EndpointA.SetChannel(channel)
+			},
+			types.ErrInvalidChannelState,
+		},
+		{
+			"channel has in-flight packets",
+			func() {
+				portID := path.EndpointA.ChannelConfig.PortID
+				channelID := path.EndpointA.ChannelID
+				// Set a dummy packet commitment to simulate in-flight packets
+				suite.chainA.GetSimApp().IBCKeeper.ChannelKeeper.SetPacketCommitment(suite.chainA.GetContext(), portID, channelID, 1, []byte("hash"))
+			},
+			types.ErrPendingInflightPackets,
+		},
+		{
+			"flush status is not FLUSHCOMPLETE",
+			func() {
+				channel := path.EndpointA.GetChannel()
+				// Set in acceptable state
+				channel.State = types.TRYUPGRADE
```

Comment:
don't particularly like having to set up the channel in a correct state but I think it's a trade-off between having the handshake be performed in the malleate function vs inside the loop body.

Reply:
My experience is that it is more maintainable to use setup handlers rather than hardcoded expectations